polygon/heimdall/simulator/bor_events.go

```diff
@@ -0,0 +1,45 @@
+package simulator
+
+import (
+	"context"
+	"time"
+
+	"github.com/ledgerwatch/erigon-lib/downloader/snaptype"
+	"github.com/ledgerwatch/erigon/polygon/heimdall"
+	"github.com/ledgerwatch/erigon/turbo/snapshotsync/freezeblocks"
+)
+
+func continueLoop(events []*heimdall.EventRecordWithTime, to time.Time, limit int) bool {
+	if len(events) == 0 {
+		return true
+	}
+	if len(events) == limit {
+		return false
+	}
+
+	last := events[len(events)-1]
+	return last.Time.Before(to)
+}
+
+func (h *HeimdallSimulator) getEvents(ctx context.Context, fromId uint64, to time.Time, limit int) ([]*heimdall.EventRecordWithTime, error) {
+	span, err := h.blockReader.EventsById(fromId, to, limit)
+
+	view := h.knownBorSnapshots.View()
+	defer view.Close()
+
+	for continueLoop(span, to, limit) && err == nil {
```

Comment:
would suggest to remove `continueLoop` func and inline the conditions in here to something roughly like `for len(events) < limit && len(events) > prevLen` - note the len(events) > prevLen indicates that EventsById(fromId, to, limit) has stopped giving us new events so we can stop as that would mean we have reached the "to"